src/optimizer/cardinality_updater.cpp

```diff
@@ -0,0 +1,137 @@
+#include "planner/operator/logical_intersect.h"
+#include "planner/operator/logical_limit.h"
+#include "planner/operator/logical_plan.h"
+
+namespace kuzu::optimizer {
+void CardinalityUpdater::rewrite(planner::LogicalPlan* plan) {
+    visitOperator(plan->getLastOperator().get());
+}
+
+void CardinalityUpdater::visitOperator(planner::LogicalOperator* op) {
+    for (auto i = 0u; i < op->getNumChildren(); ++i) {
+        visitOperator(op->getChild(i).get());
+    }
+    visitOperatorSwitchWithDefault(op);
+}
+
+void CardinalityUpdater::visitOperatorSwitchWithDefault(planner::LogicalOperator* op) {
+    switch (op->getOperatorType()) {
+    case planner::LogicalOperatorType::SCAN_NODE_TABLE: {
+        visitScanNodeTable(op);
+        break;
+    }
+    case planner::LogicalOperatorType::EXTEND: {
+        visitExtend(op);
+        break;
+    }
+    case planner::LogicalOperatorType::HASH_JOIN: {
+        visitHashJoin(op);
+        break;
+    }
+    case planner::LogicalOperatorType::CROSS_PRODUCT: {
+        visitCrossProduct(op);
+        break;
+    }
+    case planner::LogicalOperatorType::INTERSECT: {
+        visitIntersect(op);
+        break;
+    }
+    case planner::LogicalOperatorType::FLATTEN: {
+        visitFlatten(op);
+        break;
+    }
+    case planner::LogicalOperatorType::FILTER: {
+        visitFilter(op);
+        break;
+    }
+    case planner::LogicalOperatorType::LIMIT: {
+        visitLimit(op);
+        break;
+    }
+    case planner::LogicalOperatorType::AGGREGATE: {
+        visitAggregate(op);
+        break;
+    }
+    // TODO(Royi) handle the below operators as well
+    case planner::LogicalOperatorType::ACCUMULATE:
+    default: {
+        visitOperatorDefault(op);
+        break;
+    }
+    }
+}
+
+void CardinalityUpdater::visitOperatorDefault(planner::LogicalOperator* op) {
+    KU_ASSERT(op->getNumChildren() <= 1);
+    if (op->getNumChildren() == 1) {
+        op->setCardinality(op->getChild(0)->getCardinality());
+    }
+}
+
+void CardinalityUpdater::visitScanNodeTable(planner::LogicalOperator* op) {
+    op->setCardinality(cardinalityEstimator.estimateScanNode(*op));
+}
+
+void CardinalityUpdater::visitExtend(planner::LogicalOperator* op) {
+    KU_ASSERT(transaction);
+    auto& extend = op->cast<planner::LogicalExtend&>();
+    const auto extensionRate = cardinalityEstimator.getExtensionRate(*extend.getRel(),
+        *extend.getBoundNode(), transaction);
+    extend.setCardinality(cardinalityEstimator.estimateExtend(extensionRate, *op->getChild(0)));
+}
+
+void CardinalityUpdater::visitHashJoin(planner::LogicalOperator* op) {
+    auto& hashJoin = op->cast<planner::LogicalHashJoin&>();
+    KU_ASSERT(hashJoin.getNumChildren() >= 2);
+    hashJoin.setCardinality(cardinalityEstimator.estimateHashJoin(
+        hashJoin.isNodeIDOnlyJoin() ? hashJoin.getJoinNodeIDs() : binder::expression_vector{},
+        *hashJoin.getChild(0), *hashJoin.getChild(1)));
```

Comment:
In the hash joins produced by the filter push down optimizer (I think those are the only ones with non-nodeID-only joins) how do we estimate the cardinalities?

Reply:
update, for non-ID based joins for now we will go with the simplest approach of using the cross product cardinality multiplied by a magic constant